hedy.py

```diff
@@ -189,188 +212,231 @@ def list_access(self, args):
-      iterator = str(args[0])
-      commands = args[1:]
-      # in level 8 always a string
-      return [Assignment(iterator, 'string')] + self.filter_ask_assign(commands)
+    def ask(self, tree):
+        # in level 1 there is no variable name on the left side of the ask command
+        if self.level > 1:
+            self.add_to_lookup(tree.children[0].children[0], tree)
 
-    # level 9
-    def for_loop(self, args):
-      # for loop iterator is a var so should be added to the list of vars
-      iterator = str(args[0])
-      commands = args[1:]
-      # for loop iterators can be numbers (for i in range... or strings (for dier in dieren)
-      return [Assignment(iterator, 'any')] + self.filter_ask_assign(commands)
+    def input(self, tree):
+        var_name = tree.children[0].children[0]
+        self.add_to_lookup(var_name, tree)
 
+    def assign(self, tree):
+        var_name = tree.children[0].children[0]
+        self.add_to_lookup(var_name, tree.children[1])
 
-    def input(self, args):
-        # return left side of the =
-        return args[0]
+    def assign_list(self, tree):
+        var_name = tree.children[0].children[0]
+        self.add_to_lookup(var_name, tree)
 
-    def ask(self, args):
-        # try is needed cause in level 1 ask has no variable in front
-        try:
-            return Assignment(args[0], 'string')
-        except:
-            return None
+    # list access is added to the lookup table not because it must be escaped
+    # for example we print(dieren[1]) not print('dieren[1]')
+    def list_access(self, tree):
+        list_name = hash_var(tree.children[0].children[0])
+        if tree.children[1] == 'random':
+            name = f'random.choice({list_name})'
+        else:
+            # We want list access to be 1-based instead of 0-based, hence the -1
+            name = f'{list_name}[{tree.children[1]}-1]'
+        self.add_to_lookup(name, tree, True)
+
+    def list_access_var(self, tree):
+        self.add_to_lookup(tree.children[0].children[0], tree, True)
+
+    def change_list_item(self, tree):
+        self.add_to_lookup(tree.children[0].children[0], tree, True)
+
+    def repeat_list(self, tree):
+        iterator = str(tree.children[0].children[0])
+        # the tree is trimmed to skip contain the inner commands of the loop since
+        # they are not needed to infer the type of the iterator variable
+        trimmed_tree = Tree(tree.data, tree.children[0:2], tree.meta)
+        self.add_to_lookup(iterator, trimmed_tree)
+
+    def for_loop(self, tree):
+        iterator = str(tree.children[0])
+        # the tree is trimmed to skip contain the inner commands of the loop since
+        # they are not needed to infer the type of the iterator variable
+        trimmed_tree = Tree(tree.data, tree.children[0:3], tree.meta)
+        self.add_to_lookup(iterator, trimmed_tree)
+
+    def add_to_lookup(self, name, tree, skip_hashing=False):
```

Comment:
love the extra argument to indicate hashing, a lot less error-prone and more clear this way